Fix #4: Make @capture an enumerated attribute

anssiko opened the pull request:
There has been sufficient interest among implementers to beef up the `capture` hint, so here's the first stab at it.

Related issues:
https://github.com/w3c/html-media-capture/issues/4
https://github.com/whatwg/html/issues/1102

PTAL @miguelao @riju @hober

anssiko committed:
Fix #4: Make @capture an enumerated attribute

anssiko commented:
See also: [RawGit preview][1]

[1]: https://rawgit.com/w3c/html-media-capture/facingmode/index.html#the-capture-attribute

anssiko commented:
Looping in @legion80 too.

yell0wd0g commented on index.html:
```diff
@@ -178,30 +190,47 @@
+          The <dfn data-dfn-for="CaptureFacingMode">CaptureFacingMode</dfn>
+          enumeration is used to express the <a>preferred facing mode</a>. The
+          semantics of its keywords mirror the similarly named keywords defined
+          in <a><code>VideoFacingModeEnum</code></a>.
```
If it's similar to [`VideoFacingModeEnum`](https://www.w3.org/TR/mediacapture-streams/#idl-def-VideoFacingModeEnum), why not just reuse it?  I think supporting (ignoring) the `left` and `right` modes will be easier than to add another enum here.

anssiko replied:
The VideoFacingModeEnum is video-specific, so thought adding an indirection would be justified, as it'd allow us to have control over the semantics as needed.

That said, I don't feel strongly about this, we could reuse VideoFacingMode as a hint for audio or image capture too given the enum name itself is not web-facing.

anssiko replied:
@miguelao Let us know if the rationale for minting `CaptureFacingMode` enum sounds reasonable. Adding a new enum should be cheap in terms of implementation effort.

yell0wd0g replied:
TBH I think the naming `CaptureFacingMode` makes more sense than `VideoFacingModeEnum`, and we should just line up both specs. Go ahead with this change but we should file a bug there to rename it (there) and then bundle both, since they all describe the same concepts. (And also I wonder what's that `Enum` at the end of `VideoFacingModeEnum` doing there).

yell0wd0g commented on index.html:
```diff
@@ -178,30 +190,47 @@
+          "CaptureFacingMode">user</dfn> and <dfn data-dfn-for=
+          "CaptureFacingMode">environment</dfn>, which map to the respective
+          states <var>user</var> and <var>environment</var>. The <a>missing
+          value default</a> is the <var>environment</var> state. The <a>invalid
```
Is this true?  IIRC the current `<input capture>` just leaves it to the System App to decide what's the default orientation and we might want to maintain compatibility with it. If so, we should make it more evident/expectable than simply adding a description here, e.g. allow it to be [`nullable`](https://heycam.github.io/webidl/#dfn-nullable-type)?

anssiko replied:
Should made defaults implementation-specific indeed.

Can you expand the use case for nullable?

anssiko replied:
Added implementation-specific state in https://github.com/w3c/html-media-capture/pull/6/commits/5679164ddb3088c18efe42d9bce991f3e3551fe9

@miguelao PTAL to ensure this matches what you had in mind.

yell0wd0g commented on index.html:
```diff
@@ -228,36 +257,36 @@
         </li>
         <li>Or alternatively, to capture video using the device's local video
-        camera:
+        camera facing the environment:
```
Side question: is there an implementation status/ caniuse for this feature? (couldn't find any)

anssiko replied:
https://www.w3.org/2009/dap/wiki/ImplementationStatus

Thanks for the review!

yell0wd0g replied:
Also it's in the queue of caniuse (https://github.com/Fyrd/caniuse/issues/1115) waiting for some candid soul to add the entry (as a reference, this is the [`mediacapture-image`](https://github.com/Fyrd/caniuse/pull/2738)).

hober commented:
LGTM

anssiko committed:
Add implementation-specific state

Map 'missing value default' and 'invalid value default'
to the implementation-specific state.

yell0wd0g commented on index.html:
```diff
@@ -178,30 +190,53 @@
+        </p>
+        <p class="note">
+          If the user agent is unable to support the <a>preferred facing
+          mode</a>, it can fall back to the implementation-specific default
```
nit: s/can/MUST/

anssiko replied:
By convention, non-normative notes contain no RFC terms (because they're _non-normative_).

yell0wd0g reviewed (approved):
lgtm -- I wish there could be an automatic way to file bugs/issues with the browsers every time the spec gets a PR 😄  🤔

anssiko commented:
@miguelao @hober Thanks for your swift review! We're ready to land now.

foolip commented:
@miguelao, this attribute is still a boolean in Blink's IDL. Has this been shipped on any platform yet? If not then it'll be fine, but if it used anywhere in the wild already, this will become a problem.

anssiko commented:
@foolip, for the shipping status, see https://github.com/whatwg/html/issues/1102#issuecomment-214720497 and for the usage of the to-be-superseded boolean, see your UseCounters observations at https://bugs.chromium.org/p/chromium/issues/detail?id=240252#c19

TL:DR: the boolean usage has been low enough for everyone to be comfortable with this change. The known backwards incompatible change is https://github.com/whatwg/html/issues/1102#issuecomment-271983017

foolip commented:
Heh, https://codereview.chromium.org/303473004 is funny, both because I had no memory of it, and because this has already gone from enum to boolean once :)

I don't mind this change, just got a bit nervous when I saw things not lining up.

anssiko commented:
[@foolip you're involved in so many things it is no wonder you did not remember. Luckily _Google_ never forgets ;-)]

Now we're moving ahead with the implementation of the enum at https://crbug.com/698853 as you noticed already.

hober commented:
FWIW, this change has landed in trunk WebKit.

anssiko commented:
@hober, are these the right WebKit bugs to track?

* https://webkit.org/b/43239
* https://webkit.org/b/158945
